masonry/src/event_loop_runner.rs

```diff
@@ -206,8 +290,76 @@ impl ApplicationHandler<accesskit_winit::Event> for MainState<'_> {
         }
     }
 
+    // --- MARK: RENDER ---
+    fn render(&mut self, scene: Scene) {
+        let WindowState::Rendering {
+            window, surface, ..
+        } = &mut self.window
+        else {
+            tracing::warn!("Tried to render whilst suspended or before window created");
+            return;
+        };
+        let scale_factor = window.scale_factor();
+        // https://github.com/rust-windowing/winit/issues/2308
+        #[cfg(target_os = "ios")]
+        let size = window.outer_size();
+        #[cfg(not(target_os = "ios"))]
+        let size = window.inner_size();
+        let width = size.width;
+        let height = size.height;
+
+        if surface.config.width != width || surface.config.height != height {
+            self.render_cx.resize_surface(surface, width, height);
+        }
+
+        let transformed_scene = if scale_factor == 1.0 {
+            None
+        } else {
+            let mut new_scene = Scene::new();
+            new_scene.append(&scene, Some(Affine::scale(scale_factor)));
+            Some(new_scene)
+        };
+        let scene_ref = transformed_scene.as_ref().unwrap_or(&scene);
+
+        let Ok(surface_texture) = surface.surface.get_current_texture() else {
+            warn!("failed to acquire next swapchain texture");
+            return;
+        };
+        let dev_id = surface.dev_id;
+        let device = &self.render_cx.devices[dev_id].device;
+        let queue = &self.render_cx.devices[dev_id].queue;
+        let renderer_options = RendererOptions {
+            surface_format: Some(surface.format),
+            use_cpu: false,
+            antialiasing_support: AaSupport {
+                area: true,
+                msaa8: false,
+                msaa16: false,
+            },
+            num_init_threads: NonZeroUsize::new(1),
+        };
+        let render_params = RenderParams {
+            base_color: Color::BLACK,
+            width,
+            height,
+            antialiasing_method: vello::AaConfig::Area,
+        };
+        self.renderer
+            .get_or_insert_with(|| Renderer::new(device, renderer_options).unwrap())
+            .render_to_surface(device, queue, scene_ref, &surface_texture, &render_params)
```

Comment:
This unconditionally rendering to a surface here does limit the usefulness of this API somewhat.

To be clear, this is still the right API at the moment, but it does limit the use cases

Reply:
Absolutely, but just to be clear that's the existing code, I just moved it. I have a couple more changes that I was going to add in a followup pull request that expose just enough in order to run rendering from an external app (and integrate with other render passes). That method ends up just using the render root to get the scene and then handling everything else in the external render code.

This is what the external render code ends up looking like (note the vello_ext module is just some vello code that was private copy pasted into the project):
https://github.com/cfagot/space_survival/blob/main/src/xilem_render.rs

I could add those few other hooks to masonry sate now if you wanted (I mostly want this pull request to be integrated quickly though, because any change to the event loop runner file ends up being a pain to merge).

Reply:
> To be clear, this is still the right API at the moment

Was attempting to indicate that this is a non-blocking concern